### PR TITLE
[Implementacion Servicio Externo] - External service

### PR DESCRIPTION
**Pull request by AleFalcon**
## Summary
Implementatio External Services, OpenLibrary

## Trello Card
https://trello.com/c/H6MKvu0u/22-integraci%C3%B3n-api-externa

### CODE REVIEW

File: src/main/java/wolox/training/adapters/BookAdapter.java [hunk 1]
```diff
@@ -0,0 +1,50 @@
+package wolox.training.adapters;
+
+import com.fasterxml.jackson.databind.JsonNode;
+import java.util.concurrent.atomic.AtomicReference;
+import wolox.training.external.dtos.BookDTO;
+import wolox.training.models.Book;
+
+public class BookAdapter {
```
Comment (facucompe) [inline src/main/java/wolox/training/adapters/BookAdapter.java:10]: Add `@Component` so you can use `@Autowired` in the service

File: src/main/java/wolox/training/adapters/BookAdapter.java [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+
+  public BookDTO transformBookDTOToBook(String isbn, JsonNode request){
+    String title = request.path("title").textValue();
+    String subtitle = request.path("subtitle").textValue();
+    String publishDate = request.path("publish_date").textValue();
+    int numberPages = request.path("number_of_pages").intValue();
+    String[] authors = convertJsonNodeToArrayString(request.path("authors"));
+    String publishers = convertJsonNodeToString(request.path("publishers"));
+    return new BookDTO(isbn, title, subtitle, publishers, publishDate,
```
Comment (facucompe) [inline src/main/java/wolox/training/adapters/BookAdapter.java]: Instantiate the book and then use setters. Don't create a variable for every attribute

File: src/main/java/wolox/training/adapters/BookAdapter.java [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+        Integer
```
Comment (facucompe) [inline src/main/java/wolox/training/adapters/BookAdapter.java]: You first convert it from string to int in line 14.

File: src/main/java/wolox/training/external/dtos/BookDTO.java [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+      String title) {
+    this.title = title;
+  }
+
+  public String getSubtitle() {
+    return subtitle;
+  }
+
+  public void setSubtitle(
+      String subtitle) {
```
Comment (facucompe) [inline src/main/java/wolox/training/external/dtos/BookDTO.java]: Look out for unnecessary line jumps

File: src/main/java/wolox/training/controllers/ApiController.java [hunk 5]
```diff
@@ -44,4 +44,14 @@ protected User foundUser(String username, UserRepository userRepository){
         }
         return list.get();
     }
+
+    @ApiOperation(value = "Given the isbn of a book, search in DB or external service and return the book or an exception", response = User.class)
```
Comment (facucompe) [inline src/main/java/wolox/training/controllers/ApiController.java]: This should not be in the ApiController

File: src/main/java/wolox/training/controllers/BookController.java [hunk 6]
```diff
@@ -29,6 +28,8 @@
     @Autowired
     private BookRepository bookRepository;
 
+    private OpenLibraryService openLibraryService = new OpenLibraryService();
```
Comment (facucompe) [inline src/main/java/wolox/training/controllers/BookController.java]: Use `@Autowired` annotation

File: src/main/java/wolox/training/external/services/OpenLibraryService.java [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.http.ResponseEntity;
+import org.springframework.web.client.RestTemplate;
+import wolox.training.adapters.BookAdapter;
+import wolox.training.errors.book.BookHttpErrors;
+import wolox.training.external.dtos.BookDTO;
+import wolox.training.models.Book;
+import wolox.training.utils.VariablesConstants;
+
+public class OpenLibraryService {
```
Comment (facucompe) [inline src/main/java/wolox/training/external/services/OpenLibraryService.java:23]: Use `@Service` so you can use  `@Autowired` in the controller

File: src/main/java/wolox/training/utils/VariablesConstants.java [hunk 8]
```diff
@@ -8,4 +8,7 @@
 
   //VARIABLES ON USERS TEST
   public static String BOOK_URL = "/api/books/";
+
+  //EXTERNAL SERVICE
+  public static String OPEN_LIBRARY_URL = "https://openlibrary.org/api/books?bibkeys=ISBN:%s&format=json&jscmd=data";
```
Comment (facucompe) [inline src/main/java/wolox/training/utils/VariablesConstants.java]: `https://openlibrary.org/api` should be a property. Use `@Value` https://www.baeldung.com/spring-value-annotation

Comment (facucompe) [reply src/main/java/wolox/training/utils/VariablesConstants.java]: Leave `/books?bibkeys=ISBN:%s&format=json&jscmd=data` as a constant

File: src/main/java/wolox/training/external/services/OpenLibraryService.java [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+    RestTemplate restTemplate = new RestTemplate();
+    String fooResourceUrl
+        = String.format(VariablesConstants.OPEN_LIBRARY_URL, isbn);
+    ResponseEntity<String> response
+        = restTemplate.getForEntity(fooResourceUrl, String.class);
+    ObjectMapper mapper = new ObjectMapper();
+    try {
+      JsonNode root = mapper.readTree(response.getBody());
+      BookDTO bookDTO = bookAdapter.transformBookDTOToBook(isbn, root.iterator().next());
+      return bookAdapter.transformBookDTOToBook(bookDTO);
```
Comment (facucompe) [inline src/main/java/wolox/training/external/services/OpenLibraryService.java]: No need to use a  intermediate DTO here. Delete `BookDTO`

**Review by facucompe — CHANGES_REQUESTED** [review]

File: src/main/java/wolox/training/external/services/OpenLibraryService.java [hunk 10]
```diff
@@ -0,0 +1,45 @@
...
+public class OpenLibraryService {
+
+  @Autowired
+  private BookAdapter bookAdapter = new BookAdapter();
+
+  public OpenLibraryService(){
+  }
+
+  public Book bookInfo(
+      String isbn) {
```
Comment (facucompe) [inline src/main/java/wolox/training/external/services/OpenLibraryService.java]: Same  in this method. Unnecessary line jumps make code really difficult to read.

**Review by AleFalcon — COMMENTED** [review]
Ready

File: src/main/java/wolox/training/external/services/OpenLibraryService.java [hunk 11]
```diff
@@ -0,0 +1,50 @@
...
+import wolox.training.models.Book;
+import wolox.training.utils.VariablesConstants;
+
+@Service
+public class OpenLibraryService {
+
+  @Autowired
+  private BookAdapter bookAdapter;
+
+  @Value("https://openlibrary.org/api/books?bibkeys=ISBN:")
```
Comment (facucompe) [inline src/main/java/wolox/training/external/services/OpenLibraryService.java]: `@Value("${openlibrary.baseUrl}")` and add it to your application properties

**Review by facucompe — CHANGES_REQUESTED** [review]

**Review by AleFalcon — COMMENTED** [review]
Ready

**Review by facucompe — APPROVED** [review]

**Review by Gianni-Nicolas — APPROVED** [review]
Great Job

**Pull request merged**